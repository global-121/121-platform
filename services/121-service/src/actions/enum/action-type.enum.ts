export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
