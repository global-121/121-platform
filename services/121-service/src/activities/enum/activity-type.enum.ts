export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
