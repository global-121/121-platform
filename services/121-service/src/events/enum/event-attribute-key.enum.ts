export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
