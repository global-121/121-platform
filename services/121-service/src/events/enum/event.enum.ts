export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
