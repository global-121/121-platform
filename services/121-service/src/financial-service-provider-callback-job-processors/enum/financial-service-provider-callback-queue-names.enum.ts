export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
