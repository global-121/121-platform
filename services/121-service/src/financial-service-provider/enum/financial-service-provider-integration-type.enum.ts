export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
