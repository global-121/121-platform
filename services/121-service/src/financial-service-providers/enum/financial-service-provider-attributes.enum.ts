export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
