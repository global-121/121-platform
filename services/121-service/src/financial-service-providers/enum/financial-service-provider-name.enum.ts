export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
