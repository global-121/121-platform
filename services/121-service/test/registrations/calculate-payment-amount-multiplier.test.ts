import { HttpStatus } from '@nestjs/common';

import { SeedScript } from '@121-service/src/scripts/seed-script.enum';
import {
  importRegistrations,
  searchRegistrationByReferenceId,
} from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import {
  programIdPV,
  programIdWesteros,
  registrationPV5,
  registrationWesteros1,
} from '@121-service/test/registrations/pagination/pagination-data';

describe('Set/calculate payment amount multiplier', () => {
  let accessToken: string;

  beforeEach(async () => {
    accessToken = await getAccessToken();
  });

  it('should automatically calculate payment amount based on formula', async () => {
    await resetDB(SeedScript.testMultiple);
    const nrOfDragons = 2;

    // Arrange
    const registrationWesterosCopy = { ...registrationWesteros1 };
    registrationWesterosCopy.dragon = nrOfDragons;

    // Act
    const responseImport = await importRegistrations(
      programIdWesteros,
      [registrationWesterosCopy],
      accessToken,
    );

    const searchRegistrationResponse = await searchRegistrationByReferenceId(
      registrationWesterosCopy.referenceId,
      programIdWesteros,
      accessToken,
    );
    const importedRegistration = searchRegistrationResponse.body.data[0];

    // Assert
    expect(responseImport.statusCode).toBe(HttpStatus.CREATED);
    expect(importedRegistration.paymentAmountMultiplier).toBe(nrOfDragons + 1);
  });

  it('should error if paymentAmountMultiplier is set while program has a formula', async () => {
    await resetDB(SeedScript.testMultiple);
    // Arrange
    const registrationWesterosCopy = {
      ...registrationWesteros1,
      ...{ paymentAmountMultiplier: 3 },
    };

    // Act
    const responseImport = await importRegistrations(
      programIdWesteros,
      [registrationWesterosCopy],
      accessToken,
    );

    const searchRegistrationResponse = await searchRegistrationByReferenceId(
      registrationWesterosCopy.referenceId,
      programIdWesteros,
      accessToken,
    );
    // Assert
    expect(responseImport.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(searchRegistrationResponse.body).toMatchSnapshot();
  });

  it('should set paymentAmountMultiplier to 1 if program has no formula and paymentAmountMultiplier in import is not set', async () => {
    await resetDB(SeedScript.nlrcMultiple);
    // Arrange
    const registrationPvCopy = {
      ...registrationPV5,
    };

    // Act
    const responseImport = await importRegistrations(
      programIdPV,
      [registrationPvCopy],
      accessToken,
    );

    const searchRegistrationResponse = await searchRegistrationByReferenceId(
      registrationPvCopy.referenceId,
      programIdPV,
      accessToken,
    );
    const importedRegistration = searchRegistrationResponse.body.data[0];
    // Assert
    expect(responseImport.statusCode).toBe(HttpStatus.CREATED);
    expect(searchRegistrationResponse.body.data.length).toBe(1);
    expect(importedRegistration.paymentAmountMultiplier).toBe(1);
  });

  it('should set paymentAmountMultiplier based paymentAmountMultiplier if program has no formula', async () => {
    await resetDB(SeedScript.nlrcMultiple);
    // Arrange
    const paymentAmountMultiplier = 3;
    const registrationPvCopy = {
      ...registrationPV5,
      ...{ paymentAmountMultiplier },
    };
    // Act
    const responseImport = await importRegistrations(
      programIdPV,
      [registrationPvCopy],
      accessToken,
    );

    const searchRegistrationResponse = await searchRegistrationByReferenceId(
      registrationPvCopy.referenceId,
      programIdPV,
      accessToken,
    );
    const importedRegistration = searchRegistrationResponse.body.data[0];
    // Assert
    expect(responseImport.statusCode).toBe(HttpStatus.CREATED);
    expect(importedRegistration.paymentAmountMultiplier).toBe(
      paymentAmountMultiplier,
    );
  });
});
