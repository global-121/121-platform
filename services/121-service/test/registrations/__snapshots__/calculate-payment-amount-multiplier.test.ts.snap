// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set/calculate payment amount multiplier should error if paymentAmountMultiplier is set while program has a formula 1`] = `
{
  "data": [],
  "links": {
    "current": "http://localhost:3000/api/programs/2/registrations?page=1&limit=20&sortBy=id:ASC&",
  },
  "meta": {
    "currentPage": 1,
    "filter": {},
    "itemsPerPage": 20,
    "sortBy": [
      [
        "id",
        "ASC",
      ],
    ],
    "totalItems": 0,
    "totalPages": 0,
  },
}
`;
